##  Proyecto 1:
### EDA/GEDA

**Cargando librerías:**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 
from pandas import DataFrame

**Cargando archivo:**

In [2]:
df = pd.read_csv("datos/incidentes-viales-c5.csv")

**Obteniendo infromación general del dataset**

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1383138 entries, 0 to 1383137
Data columns (total 18 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   folio              1383138 non-null  object 
 1   fecha_creacion     1383138 non-null  object 
 2   hora_creacion      1383138 non-null  object 
 3   dia_semana         1383138 non-null  object 
 4   codigo_cierre      1383138 non-null  object 
 5   fecha_cierre       1383138 non-null  object 
 6   año_cierre         1383138 non-null  int64  
 7   mes_cierre         1383138 non-null  object 
 8   hora_cierre        1383138 non-null  object 
 9   delegacion_inicio  1382980 non-null  object 
 10  incidente_c4       1383138 non-null  object 
 11  latitud            1382695 non-null  float64
 12  longitud           1382703 non-null  float64
 13  clas_con_f_alarma  1383138 non-null  object 
 14  tipo_entrada       1383138 non-null  object 
 15  delegacion_cierre  1382998 non-n

**Muestra de datos:**

In [4]:
df.head(5)

folio fecha_creacion hora_creacion dia_semana  \
0  GA/160123/05714     23/01/2016      22:35:04     Sábado   
1  AO/160123/05826     23/01/2016      22:50:49     Sábado   
2  C4/160124/02187     24/01/2016      09:40:11    Domingo   
3  C4/160124/05723     24/01/2016      22:40:57    Domingo   
4  C4/160124/01334     24/01/2016      04:25:15    Domingo   

                                       codigo_cierre fecha_cierre  año_cierre  \
0  (A) La unidad de atención a emergencias fue de...   24/01/2016        2016   
1  (A) La unidad de atención a emergencias fue de...   24/01/2016        2016   
2  (N) La unidad de atención a emergencias fue de...   24/01/2016        2016   
3  (N) La unidad de atención a emergencias fue de...   24/01/2016        2016   
4  (A) La unidad de atención a emergencias fue de...   24/01/2016        2016   

  mes_cierre hora_cierre    delegacion_inicio  \
0      Enero    00:21:08  VENUSTIANO CARRANZA   
1      Enero    04:40:37           CUAJIMALPA   
2      Enero    11:13:03              TLALPAN   
3      Enero    22:55:51  MAGDALENA CONTRERAS   
4      Enero    05:03:59       MIGUEL HIDALGO   

                      incidente_c4    latitud   longitud  clas_con_f_alarma  \
0  accidente-choque sin lesionados  19.422113 -99.084643         EMERGENCIA   
1  accidente-choque con lesionados  19.358320 -99.297641  URGENCIAS MEDICAS   
2  accidente-choque sin lesionados  19.217070 -99.219070         EMERGENCIA   
3  accidente-choque sin lesionados  19.320580 -99.241010         EMERGENCIA   
4  accidente-choque sin lesionados  19.452900 -99.215870         EMERGENCIA   

       tipo_entrada    delegacion_cierre                  geopoint  mes  
0  BOTÓN DE AUXILIO  VENUSTIANO CARRANZA      19.422113,-99.084643    1  
1  BOTÓN DE AUXILIO           CUAJIMALPA       19.35832,-99.297641    1  
2   LLAMADA DEL 066              TLALPAN     19.21707,-99.21907008    1  
3   LLAMADA DEL 066  MAGDALENA CONTRERAS  19.32057999,-99.24100992    1  
4   LLAMADA DEL 066       MIGUEL HIDALGO  19.45289997,-99.21587004    1

**Resumen:**

1. ¿Cuántas variables tenemos? ***18***

2. ¿Cuántas observaciones tenemos? ***1,383,138***

3. ¿Cuántas variables numéricas tenemos? ***4 variables numéricas***  (año_cierre, latitud, longitud, mes)
    -  ¿Cambiar año y mes a categórica?

4. ¿Cuántas variables de fecha tenemos? ***2 de fecha***  (fecha_creacion, fecha_cierre)

5. ¿Cuántas variables categóricas tenemos? ***8 variables categóricas***  (dia_semana, codigo_cierre, mes_cierre, delegacion_inicio, incidente_c4, clas_con_f_alarma, tipo_entrada, delegacion_cierre)

6. ¿Cuántas variables de texto tenemos? ***4 variable de texto***  (folio, hora_creacion, hora_cierre, geopoint)
    -  ¿Eliminar geopoint?

7. ¿Cuántas observaciones únicas tenemos por variable?: 

In [5]:
def unicos_val_by_col(df):
    #Obtiene los valores únicos por columa, recibe un dataframe.
    for col in list(df):
        print(col + ": " , len(df[col].unique().tolist()))
        
unicos_val_by_col(df)

folio:  1383138
fecha_creacion:  2637
hora_creacion:  105887
dia_semana:  7
codigo_cierre:  5
fecha_cierre:  2621
año_cierre:  7
mes_cierre:  12
hora_cierre:  106827
delegacion_inicio:  17
incidente_c4:  26
latitud:  82502
longitud:  78985
clas_con_f_alarma:  4
tipo_entrada:  9
delegacion_cierre:  17
geopoint:  134289
mes:  12


**Observaciones:**

-  Hay 1,383,138 folios.
-  Hay datos de 7 años (2014 - 2020).
-  Hay 17 delegaciones en vez de 16, tanto de inicio como de cierre.
    -  Sí son 16, ¿por qué la función indica 17?
-  Sí coinciden los códigos de cierre (5) y las clasificaciones (4) con el diccionario.
-  Hay 26 tipos de incidente, contra 25 que muestra el diccionario.
-  Hay 9 tipos de entrada, contra 7 que muestra el diccionario.

###  Profiling inicial:

In [6]:
def describe_fast(df, p):
    #Esta funcion arroja el resumen de información por tipo de variable
    # El valor p es 1 para variables categoricas, 0 para variables numericas
    
    cat = df.select_dtypes(np.object)
    num = df.select_dtypes(np.number)
    n = len(cat.columns)
    m = len(num.columns)
    
    if p== 1: 
        
        for i in range(0,n):
            print("Variable:", cat.columns[i])
            print("Número de registros únicos/categorías:",len(cat.iloc[:,i].unique())) 
            print("Porcentaje de valores nulos:", sum(pd.isna(cat.iloc[:,1]))/len(cat.iloc[:,1]))
            print("Distribución:")
            print(cat.iloc[:,i].value_counts()/ len(cat.iloc[:,i]),"\n")
    else:
        for i in range(0,m):
            print(num.iloc[:,i].describe(),)
            print("Número de registros únicos:", len(num.iloc[:,i].unique()))
            print("Porcentaje de valores nulos:", sum(pd.isna(num.iloc[:,i]))/len(num.iloc[:,i]),"\n")

**Variables numéricas:**

In [7]:
describe_fast(df, 2)

count    1.383138e+06
mean     2.017023e+03
std      1.869989e+00
min      2.014000e+03
25%      2.015000e+03
50%      2.017000e+03
75%      2.019000e+03
max      2.020000e+03
Name: año_cierre, dtype: float64
Número de registros únicos: 7
Porcentaje de valores nulos: 0.0 

count    1.382695e+06
mean     1.938391e+01
std      2.666382e-01
min      1.909402e+01
25%      1.933685e+01
50%      1.938408e+01
75%      1.943496e+01
max      1.953030e+02
Name: latitud, dtype: float64
Número de registros únicos: 82502
Porcentaje de valores nulos: 0.0003202861898089706 

count    1.382703e+06
mean    -9.914359e+01
std      2.399677e+00
min     -9.917640e+02
25%     -9.917930e+01
50%     -9.914023e+01
75%     -9.909600e+01
max     -9.894537e+01
Name: longitud, dtype: float64
Número de registros únicos: 78985
Porcentaje de valores nulos: 0.00031450224055734133 

count    1.383138e+06
mean     6.617056e+00
std      3.460213e+00
min      1.000000e+00
25%      4.000000e+00
50%      7.000000e+00
75%   

**Observaciones:**

-  Los años van del 2014 al 2020 (7 en total).
-  ¿Cuáles son los años con más folios totales?
-  ¿Cuáles son los años con más folios falsos?
-  ¿Cuáles son los meses con más folios totales?
-  ¿Cuáles son los meses con más folios falsos?
-  Mapa con folios y distribución de llamadas falsas y verdaderas (latitud y longitud).

**Variables categóricas:**

In [9]:
describe_fast(df, 1)

Variable: folio
Número de registros únicos/categorías: 1383138
Porcentaje de valores nulos: 0.0
Distribución:
C4/141007/01165    7.229937e-07
C4/150611/00945    7.229937e-07
C5/161122/03371    7.229937e-07
C5/191006/08209    7.229937e-07
C5/180728/00574    7.229937e-07
                       ...     
C5/190521/08519    7.229937e-07
C5/170703/01579    7.229937e-07
C5/180721/06550    7.229937e-07
C5/181124/00769    7.229937e-07
C5/190415/07143    7.229937e-07
Name: folio, Length: 1383138, dtype: float64 

Variable: fecha_creacion
Número de registros únicos/categorías: 2637
Porcentaje de valores nulos: 0.0
Distribución:
14/02/2020    8.408416e-04
26/10/2018    7.432375e-04
08/12/2017    7.309466e-04
30/11/2019    7.309466e-04
21/04/2018    7.295006e-04
                  ...     
04/03/2019    7.229937e-07
13/07/2019    7.229937e-07
06/08/2019    7.229937e-07
21/03/2019    7.229937e-07
27/08/2019    7.229937e-07
Name: fecha_creacion, Length: 2637, dtype: float64 

Variable: hora_creacion
N

Número de registros únicos/categorías: 134289
Porcentaje de valores nulos: 0.0
Distribución:
19.37168001,-99.08024004    1.563112e-03
19.30431996,-99.08714016    1.513949e-03
19.347021,-99.180646        1.250779e-03
19.35663003,-99.08655984    1.082322e-03
19.36316997,-99.05739984    1.017252e-03
                                ...     
19.56454,-99.13759          7.229937e-07
19.19197845,-99.0226593     7.229937e-07
19.28632791,-99.23747174    7.229937e-07
19.46759,-99.09793          7.229937e-07
19.34632833,-99.17717       7.229937e-07
Name: geopoint, Length: 134288, dtype: float64 



**Observaciones:**

-  Las llamadas falsas (N + F) equivalen al 20.3% de las llamadas totales.
-  El día con más llamadas es el 14/feb/2020.  
    -  ¿Hay alguna fecha con más proporción de llamadas falsas?
    -  ¿Cuál es la distribución de llamadas falsas y verdaderas por año, mes, día de la semana, y hora?
-  La delegación con más llamadas es Iztapalapa (16%), Milpa Alta es la que menos tiene (0.05%).    
    -  ¿Cuál es la distribución de llamadas falsas y verdaderas para cada delegación?
-  ¿Nos interesa la relación (tiempo y/o delegación), entre la creación y el cierre?
-  El tipo de incidente (incidente_c4), ¿es un dato de entrada o de salida?  
    -  ¿Cuál es la proporción de llamadas falsas?
-  La clasificación (clas_con_f_alarma), ¿es un dato de entrada o de salida?
    -  ¿Cuál es la proporción de llamadas falsas?
-  El tipo de entrada más común es "Llamada del 911" (53.4%), el menos común es "Aplicativos" (0.006%)
    -  ¿Cuál es la distribución de llamadas falsas y verdaderas para cada tipo de entrada?
-  Hay 16 delegaciones, ¿el profiling dice que son 17?
-  ¿Hay ubicaciones específicas con un alto índice de llamadas falsas?

### Gráficas:

-  Mapa con llamadas falsas y verdaderas.  (O)
    -  ¿Zoom en el mapa con alguna zona de alta incidencia de llamadas falsas?
-  Gráfica de barras con proporción de la etiqueta (llamadas falsas y verdaderas).  (O)
-  Gráfica con proporción de llamadas falsas y verdaderas por año.
-  Gráfica con proporción de llamadas falsas y verdaderas por mes.
-  Gráfica con proporción de llamadas falsas y verdaderas por día de la semana.
-  Gráfica con proporción de llamadas falsas y verdaderas por cada 3 horas.
-  Gráfica con proporción de llamadas falsas y verdaderas por cada tipo de entrada.
-  Gráfica con proporción de llamadas falsas y verdaderas por delegación.